docker/Dockerfile.base

```diff
@@ -28,6 +28,10 @@ RUN apt-get update && \
    && apt-get clean \
    && rm -rf  /var/log/*log /var/lib/apt/lists/* /var/log/apt/* /var/lib/dpkg/*-old /var/cache/debconf/*-old
 
+# Create nonroot user that aligns with distroless uid/gid/path
+RUN groupadd --gid 65532 nonroot && \
```

Comment:
Why not run as UID 1337 for consistency? I think all other control plane components already run as non-root and don't need this.

Reply:
65532 is the default UID/GID for nonroot users in distroless. I thought it more appropriate to align with distroless, rather than 1337. In a distroless world, I am not sure how we (easily) introduce a 1337 user, whereas nonroot is ready to go and USER defaults to this uId.

Reply:
to add clarity - AFAIK 1317 is not available in a distroless build. This is why I added the nonroot UID/GID/path - that is equivalent to that used in the bezel files in distroless.

Reply:
I am not 100% sure of  all the details, but we run Pilot as non-root right? Based on that, and the fact we do not already have this stuff, I would expect we can do it another way?

If this way is better, we should also switch pilot/others over to this way as well - I don't really care how we do it (because I don't know best practices here) just that its consistent

Reply:
@howardjohn switching pilot is a straightforward sed 1337/65532. The problem arises in that I am not sure what the upgrade impact looks like. The iptables code all over the codebase expects 1337 as the PID - however - once we move to distroless 1337 will no longer be present.

Reply:
Pilot does not have iptables rules applied. Are you sure 1337 is not present on distroless? I don't see how distroless can possibly function if so (in the proxy)

Reply:
POSIX requires any UID/GID that is in use to be readable by the glibc implementation. How this is done is an exercise to the reader. Many glibc calls (such as getpwent() rely on a valid home directory, a valid uid entry, and a valid GID entry in /etc/password. The discovery process does not appear to need this, although the operator process does appear to need this.

I could not get the operator to run properly unless it was running as root, or I added the above UID/GID/home dir to the system. Otherwise, the operator craters when it executes the getpwent() system call.

"nonroot" is present in distroless already. I had thought a simple approach would be to just use the nonroot user for both distroless and non-distroless. If you wish, I can use nonroot for distroless, and "istio-proxy" for ubuntu based distributions.

This leaves open the question of how is a password file/groups file/shadow file generated in a world where distroless is the only distribution, and 1337 is required and baked into the implementation.z

Unfortunately, we need 2 UIDs - as 1337 is used by iptables in the proxy. 

A few alternatives for distroless exist to creating the home dir/uid entry/gid entry. A multibuild of distroless may work:
* Build a throw-away container as step 1 which copies /etc/passwd, /etc/shadow, /etc/groups to some location (such as /home).
* Modify /home/passwd, /home/shadow, /home/groups to add 1337
* Copy /home/passwd, /home/shadow, /home/groups to the distroless container during step 2 of the build (the one that actually builds the distroless container.

At the moment, if the project strives for consistency, the above solution will need to be implemented.

Open to other ideas.....

Reply:
Should we just build our own distroless images? See https://github.com/istio/istio/issues/17656, we already do basically a hackier version of your suggestion with a throw away container

Reply:
Yes, I have thought of roll our own distroless. It really isn't all that difficult. Another option is to pull master of distroless - apply a patch - and build using the upstream distroless code. I really hate to re-invent that wheel.

One challenge with either approach is the need for bezel. Apparently bezel does not work well in nested virt (bezel uses its own virt layer) - and some peeps use virtualization to build or test Istio.

Cheers,
-steve

Reply:
base image for distroless: https://github.com/istio/istio/pull/24929